compiler/rustc_mir/src/transform/inline.rs

```diff
@@ -115,7 +124,10 @@ impl Inliner<'tcx> {
                     // in the other direction.
                     if !self.tcx.dep_graph.is_fully_enabled()
                         && self_hir_id < callee_hir_id
```

Comment:
It's not needed, this is just a fast path

Reply:
This check prevents inlining in about half of the cases.

Reply:
This check is just a way to not call `mir_inliner_cycle` in the cases where we know for sure that no cycle can happen. It does not prevent any inlining

Reply:
feels like `!self_recurses` have the same purpose as `!self.tcx.dep_graph.is_fully_enabled() && self_hir_id < callee_hir_id` and as `self_recurses` is already computed so that is the fastes check

Reply:
This check allows inlining even if the current function is recursive *and* the function we are calling is recursive. In all other cases, you are correct. We could remove this check, the benefit of inlining a few functions depending on a rather random information is likely very small and probably just makes things more confusing